ports/litex/boards/orangecrab/orangecrab-spi.ld

```diff
@@ -0,0 +1,116 @@
+}
+
+/* top end of the stack */
+_estack = ORIGIN(RAM) + LENGTH(RAM);
+
+/* define output sections */
+SECTIONS
+{
+    /* This is the initialized data section
+    The program executes knowing that the data is in the RAM
+    but the loader puts the initial values in the FLASH (inidata).
+    It is one task of the startup to copy the initial values from FLASH to RAM. */
+    .data :  AT ( _sidata )
+    {
+        . = ALIGN(4);
+        _sdata = .;        /* create a global symbol at data start; used by startup code in order to initialise the .data section in RAM */
+
+        *(.itcm.*)         /* Instruction Tightly Coupled Memory */
+        *(.dtcm_data.*)    /* Data Tightly Coupled Memory */
+
+        *(.text.cmp_lfn)
+        *(.text.qstr_find_strn)
+        *(.text.dcd_edpt_xfer)
+        *(.text.pop_rule)
+        *(.text.ff_wtoupper)
+        *(.text.dir_find)
+        *(.text.push_rule)
+        *(.text.csr_writel)
+        *(.text.csr_readl)
+        *(.text.timer0_ev_pending_write)
+        *(.text.autoreload_tick)
+        *(.text.filesystem_tick)
+        *(.text.usb_background)
+
+        *(.text.dcd_*)
+        *(.text.tud_control_*)
+        *(.text.tud_cdc_n_write_flush)
+        *(.text.tud_task)
+        *(.text.tu_edpt_dir)
+        *(.text.tu_fifo_empty)
+        *(.text.usbd_edpt_busy)
+        *(.text.irq_getmask)
+        *(.text.irq_setmask)
+        *(.text.irq_pending)
+        *(.text._osal_q_lock)
+        *(.text.osal_queue_receive)
+
+        *(.text.mp_obj_get_type)
+        *(.text.mp_parse)
+        *(.text.parse_compile_execute)
+        *(.text.mp_map_lookup)
+        *(.text.mp_execute_bytecode) /* Note: this function is 7kb */
+
+        *(.ramtext)        /* .text* sections (code) */
+        *(.ramtext*)       /* .text* sections (code) */
+        *(.data)           /* .data sections */
+        *(.data*)          /* .data* sections */
+        *(.sdata)          /* .data sections */
+        *(.sdata*)         /* .data* sections */
+
+        . = ALIGN(4);
+        _edata = .;        /* define a global symbol at data end; used by startup code in order to initialise the .data section in RAM */
+    } >RAM
+
+    /* The program code and other data goes into FLASH */
+    .text :
+    {
+        . = ALIGN(4);
+        KEEP(*(.text.start)) /* isr vector table */
+        *(.text)           /* .text sections (code) */
+        *(.text*)          /* .text* sections (code) */
+        *(.rodata)         /* .rodata sections (constants, strings, etc.) */
+        *(.rodata*)        /* .rodata* sections (constants, strings, etc.) */
+        *(.srodata)        /* .rodata sections (constants, strings, etc.) */
+        *(.srodata*)       /* .rodata* sections (constants, strings, etc.) */
+
+        . = ALIGN(4);
+        _etext = .;        /* define a global symbol at end of code */
+        _sidata = _etext;  /* This is used by the startup in order to initialize the .data secion */
+    } >FLASH_FIRMWARE
```

Comment:
Does all code get moved into RAM? How is that done?

Reply:
Currently that is handled by the LiteX bios. On startup the following things happen:
- SoC starts executing at 0x00000000, this memory is frimware embedded in the gateware (litex bios).
- We configure and validate that the DDR3 controller and memory is running correctly
- We look in FLASH for a RISCV firmware. 0x20100000 the firmware is prefixed with a crc32 and length.
- If crc check passes the firmware is copied into DDR3 memory address space, 0x40000000.
- We jump to 0x40000000.

It is possible to set this up in multiple ways. In this method the RISCV firmware just sees an entry point of 0x40000000, and doesn't have to worry about configuring the DDR3 memory or copying itself into RAM.

Reply:
Ok, cool! I wonder if we don't need .data in RAM then since "flash" is actually RAM. Is it possible to restart the CPU without this happening? In that case we would want a clean copy of .data in the "flash" region of RAM.

Reply:
On the Fomu, there was a register that let you set the CPU reset address. 
So that if you trigger a CPU reset, the PC is reset to 0x40000000, this way the bootloader wouldn't run again. 
The standard C startup code should be copying a fresh .data section across before calling main().

Reply:
Ok, so we need .data to be preserved. All good!